arch/arm/soc/nordic_nrf/nrf52/soc.c

```diff
@@ -382,53 +44,11 @@ static int nordicsemi_nrf52_init(struct device *arg)
 
 	key = irq_lock();
 
-#ifdef CONFIG_SOC_NRF52832
-	nordicsemi_nrf52832_init();
-#endif
-#ifdef CONFIG_SOC_NRF52840
-	nordicsemi_nrf52840_init();
-#endif
+	SystemInit();
 
 #ifdef CONFIG_NRF_ENABLE_ICACHE
 	/* Enable the instruction cache */
 	NRF_NVMC->ICACHECNF = NVMC_ICACHECNF_CACHEEN_Msk;
-#endif /* CONFIG_NRF_ENABLE_ICACHE */
-
-	/* Enable the FPU if the compiler used floating point unit
-	 * instructions. Since the FPU consumes energy, remember to
-	 * disable FPU use in the compiler if floating point unit
-	 * operations are not used in your code.
-	 */
-#if defined(CONFIG_FLOAT)
```

Comment:
What about this one? Don't we need this here?

Reply:
It is done in SystemInit(): [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/nordic/nrfx/mdk/system_nrf52.c#L175) for nRF52832 and [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/nordic/nrfx/mdk/system_nrf52840.c#L144) for nRF52840.

Reply:
`__FPU_USED` is a CMSIS or GCC thing I believe. Are we sure this is enabled when `CONFIG_FLOAT` is set?

Reply:
It's from CMSIS. I checked with `#warning` placed inside `#if (__FPU_USED == 1)` and it is set properly when CONFIG_FLOAT is active.
Then I tried the "fp_sharing" example and it ran fine. However, it apparently ran fine even if I commented out this enabling FPU part...
After some debugging I found that the FPU was enabled [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/arm/core/cortex_m/prep_c.c#L62).

Reply:
well, SystemInit() is a kind of generic function and imo should contain FPU configuration.

Reply:
Thus, we'll need some mechanism to prevent FPU initialization in `SystemInit()`, something like:
```
#if (__FPU_USED == 1) && !defined(NO_DEFAULT_FPU_INIT)
...
```
and define `NO_DEFAULT_FPU_INIT` in the build system when `CONFIG_FLOAT` is not enabled.
But `SystemInit()` should be modified in MDK and then imported (via nrfx) to Zephyr.

Issue 2. could be solved even now by adding:
```
#if (__FPU_USED == 1) && !defined(CONFIG_FLOAT)
	SCB->CPACR &= ~((3UL << 20) | (3UL << 22));
	__DSB();
	__ISB();
#endif
```
in "soc.c" right after the call to `SystemInit()`, but this would be ugly.

Can we merge it as it is now and create (I'll do it) a ticket for MDK guys to add the above option?

Reply:
Yes, let's merge as is. 

`

> #if (__FPU_USED == 1) && !defined(NO_DEFAULT_FPU_INIT)

`
this is a good solution, IMO, i.e. have an option in nRFx that instructs SystemInit() to suppress FPU initialization.

Reply:
If I read your comments correctly, there is an assumption that __FPU_USED can be 1 while the FPU is unused? To reiterate @anangl's comment above, __FPU_USED is not __FPU_PRESENT. __FPU_USED should only be 1  when the FPU is actually in use.

Reply:
@simtind The whole point is that we want to control the initialization of FPU entirely with the CONFIG_FLOAT option and to avoid this second FPU initialization (done in SystemInit() regardless of the CONFIG_FLOAT state). But still, we want to call the standard SystemInit() to have all the workarounds applied properly.